src/group_theory/subsemigroup/operations.lean

```diff
@@ -0,0 +1,696 @@
+/-
+Copyright (c) 2018 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Kenny Lau, Johan Commelin, Mario Carneiro, Kevin Buzzard,
+Amelia Livingston, Yury Kudryashov
+-/
+
+import group_theory.subsemigroup.basic
+import data.equiv.mul_add
+import algebra.group.prod
+import algebra.group.inj_surj
+import group_theory.group_action.defs
+
+/-!
+# Operations on `subsemigroup`s
+
+In this file we define various operations on `subsemigroup`s and `mul_hom`s.
+
+## Main definitions
+
+### Conversion between multiplicative and additive definitions
+
+* `subsemigroup.to_add_subsemigroup`, `subsemigroup.to_add_subsemigroup'`,
+  `add_subsemigroup.to_subsemigroup`, `add_subsemigroup.to_subsemigroup'`:
+  convert between multiplicative and additive subsemigroups of `M`,
+  `multiplicative M`, and `additive M`. These are stated as `order_iso`s.
+
+### (Commutative) monoid structure on a subsemigroup
+
+* `subsemigroup.to_monoid`, `subsemigroup.to_comm_monoid`:
+  a subsemigroup inherits a (commutative) semigroup structure.
+
+### Group actions by subsemigroups
+
+* `subsemigroup.mul_action`, `subsemigroup.distrib_mul_action`:
+  a subsemigroup inherits (distributive) multiplicative actions.
+
+### Operations on subsemigroups
+
+* `subsemigroup.comap`: preimage of a subsemigroup under a mul homomorphism as
+  a subsemigroup of the domain;
+* `subsemigroup.map`: image of a subsemigroup under a mul homomorphism as
+  a subsemigroup of the codomain;
+* `subsemigroup.prod`: product of two subsemigroups `s : subsemigroup M` and `t : subsemigroup N`
+  as a subsemigroup of `M × N`;
+
+### Mul homomorphisms between subsemigroup
+
+* `subsemigroup.subtype`: embedding of a subsemigroup into the ambient magma.
+* `subsemigroup.inclusion`: given two subsemigroups `S`, `T` such that `S ≤ T`, `S.inclusion T`
+  is the inclusion of `S` into `T` as a monoid homomorphism;
```

Comment:
Similar to Riccardo's comment, should this say mul hom at the end? (and similarly in lines 58-62)